### ISSUE TITLE
When mongoose.Error.ValidatorError.kind === undefined

### ISSUE THREAD

**Issue by jlchereau** [Automattic/mongoose#2885]
I am in the process of upgrading a project to mongoose 4.

```type``` has been replaced by ```kind``` in  mongoose.Error.ValidatorError:
- where the value was ```"required"```, it is still ```"required"```,
- but where the value was ```"user defined"```, it is now either ```"user defined"``` or ```undefined```.

This change impacts mocha tests for custom validators where our tests are like:

```js
expect(error).to.be.an.instanceof(mongoose.Error.ValidationError).with.property('errors');
//expect(error.errors).to.have.property('address').that.is.an.instanceof(mongoose.Error.ValidatorError).with.property('kind', 'user defined');
expect(error.errors).to.have.property('address').that.is.an.instanceof(mongoose.Error.ValidatorError).with.property('kind', undefined);
expect(Object.keys(error.errors)).to.be.an.instanceof(Array).with.property('length', 1);
```

it seems that (TBC):
- this.invalidate in pre-validate middleware yields ```"user defined"```
- schema validators like {validator: **regex**, msg: '...'} yield ```undefined```
- schema validators like {validator: **function**, msg: '...'} yield ```"user defined"```

Is this expected?

I could not find in the documentation an explanation when to expect ```"user defined"``` and when to expect ```undefined```. Please enlighten me.

**Comment by vkarpov15** [comment]
The "type" -> "kind" change is documented in the [4.0 release notes list of backwards breaking changes](https://github.com/Automattic/mongoose/wiki/4.0-Release-Notes#backwards-breaking-changes). The change of the regex validator's kind is unexpected, I'll fix that.

General idea is that "user defined" is for custom validators and the built-in validator name for built-in validators (I think) so ideally validator.kind should be 'regex' for regex validator. Thanks for pointing this out.

**Comment by jlchereau** [comment]
After more checks, the mix between ```"user defined"``` and ```undefined``` is less obvious. In particular you can have test1 and test2 which yield ```undefine``` when one is triggered without the other but yield ```"user defined"``` when triggered together.

**Comment by jlchereau** [comment]
This needs to be reopened as it is not yet solved in 4.0.3.

See my comment on https://github.com/Automattic/mongoose/commit/f7f41ab8df2fff7e630b2392fd3087b746981f6f.

```
var commentSchema = new ActivityBaseSchema({
    text    : { type: String, required: true, validate: schemaValidators.longDescription }
    //Consider the option to publish comments to support forums to seek for help/assistance
});
```

and

```
longDescription: {
    validator: /^[^\s<>{}][^<>{}]{4,998}[^\s<>]$/, //Consider quotes
    msg: '`{PATH}` is invalid: `{VALUE}` is not a trimmed string between 5 and 1000 characters long or contains `<` or `>` symbols.'
},
```

still returns kind undefined whereas 3.8 returned type 'user defined'

**Comment by vkarpov15** [comment]
Thanks, sorry. This is what I get for not TDD-ing.
